analytics/transform.go

```diff
@@ -191,22 +192,116 @@ func (event ParsedEvent) GetValue(field string) (interface{}, error) {
 		return nil, errors.New(fmt.Sprintf("Key %s not a valid atomic field", field))
 	}
 	if event[index] == "" {
-		return nil, errors.New(fmt.Sprintf("Field %s is empty", field))
+		return nil, errors.New(EmptyFieldErr)
 	}
 	kvPairs, err := enrichedEventFieldTypes[index].ParseFunction(enrichedEventFieldTypes[index].Key, event[index])
 	if err != nil {
 		return nil, err
 	}
-	if field == "contexts" || field == "derived_contexts" || field == "unstruct_event" {
-		// TODO: DRY HERE TOO?
-		output := make(map[string]interface{})
+
+	return kvPairs, nil
+}
+
+// GetValue returns the value for a provided atomic field, without processing the rest of the event.
+// For unstruct_event, it returns a map of only the data for the unstruct event.
+func (event ParsedEvent) GetValue(field string) (interface{}, error) {
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+	return kvPairs[0].Value, nil
+}
+
+// GetUnstructEventValue returns the value for a provided atomic field inside an event's unstruct_event field
+func (event ParsedEvent) GetUnstructEventValue(path ...interface{}) (interface{}, error) {
+	field := "unstruct_event"
+
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the key/value pairs of the event field into a map
+	eventMap := make(map[string]interface{})
+	for _, pair := range kvPairs {
+		eventMap[pair.Key] = pair.Value
+	}
+
+	j, err := json.Marshal(eventMap)
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]interface{}, len(path))
+	for idx := range path {
+		b[idx] = path[idx]
+	}
+
+	el := json.Get(j, b...)
```

Comment:
I think the issue description and perhaps some of our brief conversations about this may have been misleading.

The idea behind suggesting the use of `json.Get()` was originally that before parsing this field, we have a [self-describing JSON](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/self-describing-jsons/). So, if we can figure out the pathing, jsoniter's `Get()` function could be used to grab the field from the original TSV value, before running the parseUnstruct (which is fairly heavy duty since it unmarshals JSON).

I don't want to block a release on this since it will work, I'm just not a fan of the unnecessary addition to performance. @TiganeteaRobert why don't we jump on a call to chat about it? Maybe there's a compromise to timebox a task to evaluate how easy the original suggestion would be to do (I have no idea btw) or something like that.

Reply:
As per our call, the GetUnstructEventValue can be greatly improved by skipping parsing and retrieving the data.data object directly from the event.